File: 1.模法推荐队-可运行项目成果/backend/app/schemas/recommendation.py
"""Schema for model recommendation."""

from __future__ import annotations
from typing import Any, Literal
from pydantic import BaseModel, ConfigDict, Field


class ScoreBreakdown(BaseModel):
    """Multi-dimensional score breakdown."""
    scenario_match: float = 0.0
    customer_match: float = 0.0
    data_match: float = 0.0
    output_match: float = 0.0
    graph_path_match: float = 0.0
    field_compatibility: float = 0.0
    hybrid_retrieval_match: float = 0.0
    llm_semantic_match: float = 0.0
    performance: float = 0.0
    landing_experience: float = 0.0
    compliance: float = 0.0


class EvidenceCard(BaseModel):
    """A single evidence item supporting the recommendation."""
    evidence_type: str = ""
    content: str = ""
    source: str = ""
    evidence_text: str = ""
    source_field: str = ""
    confidence: float = 0.0


class AlternativeModel(BaseModel):
    """Alternative model information."""
    model_id: str = ""
    model_name: str = ""
    reason: str = ""


class DataReadinessReport(BaseModel):
    """Data readiness diagnosis for a recommended model."""
    readiness_score: float = 0.0
    required_fields: list[str] = Field(default_factory=list)
    available_fields: list[str] = Field(default_factory=list)
    missing_required_fields: list[str] = Field(default_factory=list)
    missing_optional_fields: list[str] = Field(default_factory=list)
    confidence_impact: str = ""
    action_items: list[str] = Field(default_factory=list)
    substitution_notes: list[str] = Field(default_factory=list)


class RecommendedModel(BaseModel):
    """A single recommended model with full detail."""
    model_id: str = ""
    model_name: str = ""
    source: str = ""
    catalog_version: str = ""
    rank: int = 0
    total_score: float = 0.0
    rule_score: float = 0.0
    graph_score: float = 0.0
    retrieval_score: float = 0.0
    llm_score: float = 0.0
    score_breakdown: ScoreBreakdown = Field(default_factory=ScoreBreakdown)
    recommendation_reason: str = ""
    evidence_cards: list[EvidenceCard] = Field(default_factory=list)
    required_data: list[str] = Field(default_factory=list)
    missing_data: list[str] = Field(default_factory=list)
    output_fields: list[str] = Field(default_factory=list)
    applicable_boundary: str = ""
    unsuitable_conditions: str = ""
    compliance_notes: str = ""
    alternative_models: list[AlternativeModel] = Field(default_factory=list)
    data_readiness: DataReadinessReport = Field(default_factory=DataReadinessReport)


class UnrecommendedExample(BaseModel):
    """Example of a model that was not recommended and why."""
    model_id: str = ""
    model_name: str = ""
    reason: str = ""


class RecommendModelsRequest(BaseModel):
    """Request for model recommendation."""

    model_config = ConfigDict(extra="forbid")

    parse_result: dict[str, Any] = Field(description="需求解析结果")
    model_source: Literal["official", "demo", "official_then_demo"] = Field(
        default="official_then_demo",
        description="候选模型策略；默认返回官方主榜，并附带独立标注的 demo 参考候选",
    )
    top_k: int = Field(default=5, ge=1, le=20, description="返回推荐数量")
    demo_top_k: int = Field(
        default=3,
        ge=0,
        le=10,
        description="official_then_demo 策略下附带的 demo 参考候选数量",
    )
    prefer_api_available: bool = Field(default=False, description="是否优先推荐可 API 调用的模型")
    prefer_landing_cases: bool = Field(default=False, description="是否优先推荐有落地案例的模型")
    client_request_id: str = Field(
        default="",
        max_length=128,
        description="客户端幂等键；同一请求重试时保持不变",
    )


class RecommendModelsResponse(BaseModel):
    """Response containing model recommendations."""
    request_id: str = ""
    recommendations: list[RecommendedModel] = Field(default_factory=list)
    demo_references: list[RecommendedModel] = Field(default_factory=list)
    unrecommended_examples: list[UnrecommendedExample] = Field(default_factory=list)
    summary: str = ""
    catalog_policy: Literal["official", "demo", "official_then_demo"] = "official"
    demo_reference_status: Literal["not_requested", "available", "unavailable"] = "not_requested"
    official_recommendation_count: int = 0
    demo_reference_count: int = 0
    version_id: str = ""
    version_number: int = 0

File: 1.模法推荐队-可运行项目成果/frontend/src/types/recommendation.ts
import type { ParseDemandResponse } from './demand';

/** 模型推荐响应 */
export interface RecommendModelsRequest {
  parse_result: ParseDemandResponse;
  model_source?: 'official' | 'demo' | 'official_then_demo';
  top_k?: number;
  demo_top_k?: number;
  prefer_api_available?: boolean;
  prefer_landing_cases?: boolean;
  client_request_id?: string;
}

export interface RecommendModelsResponse {
  request_id: string;
  recommendations: ModelRecommendation[];
  demo_references: ModelRecommendation[];
  unrecommended_examples: UnrecommendedExample[];
  summary: string;
  catalog_policy: 'official' | 'demo' | 'official_then_demo';
  demo_reference_status: 'not_requested' | 'available' | 'unavailable';
  official_recommendation_count: number;
  demo_reference_count: number;
  version_id?: string;
  version_number?: number;
}

export interface ScoreBreakdown {
  scenario_match: number;
  customer_match: number;
  data_match: number;
  output_match: number;
  graph_path_match?: number;
  field_compatibility?: number;
  hybrid_retrieval_match?: number;
  llm_semantic_match?: number;
  performance: number;
  landing_experience: number;
  compliance: number;
}

export interface EvidenceCard {
  evidence_type: string;
  content?: string;
  source?: string;
  evidence_text?: string;
  source_field?: string;
  confidence?: number;
}

export interface AlternativeModel {
  model_id: string;
  model_name: string;
  reason?: string;
  alternative_reason: string;
  weakness_dimensions: string[];
}

export interface DataReadinessReport {
  readiness_score: number;
  required_fields: string[];
  available_fields: string[];
  missing_required_fields: string[];
  missing_optional_fields: string[];
  confidence_impact: string;
  action_items: string[];
  substitution_notes: string[];
}

export interface ModelRecommendation {
  model_id: string;
  model_name: string;
  source?: 'official' | 'demo' | string;
  catalog_version?: string;
  rank: number;
  total_score: number;
  rule_score?: number;
  graph_score?: number;
  retrieval_score?: number;
  llm_score?: number;
  score_breakdown: ScoreBreakdown;
  recommendation_reason: string;
  evidence_cards: EvidenceCard[];
  required_data: string[];
  missing_data: string[];
  output_fields: string[];
  applicable_boundary: string;
  unsuitable_conditions: string;
  compliance_notes: string;
  alternative_models: AlternativeModel[];
  data_readiness?: DataReadinessReport;
}

export interface UnrecommendedExample {
  model_id: string;
  model_name: string;
  reason: string;
}

export interface EffectEstimate {
  estimated_lift_pct: number;
  coverage_pct: number;
  confidence_band_pct: number[];
  data_readiness_factor: number;
  segment_match_factor: number;
  basis: string[];
  disclaimer: string;
  metric_source?: 'verified' | 'draft' | 'missing';
  verification_status?: string;
  evidence_level?: 'high' | 'medium' | 'low';
  assumptions?: string[];
  not_for_decision?: boolean;
}

export interface ModelComparisonItem {
  model_id: string;
  model_name: string;
  domain: string;
  customer_segment: string[];
  input_fields_required: string[];
  output_fields: string[];
  performance_metrics: Record<string, unknown>;
  applicable_conditions: string;
  unsuitable_conditions: string;
  compliance_boundary: string;
  data_readiness: DataReadinessReport;
  effect_estimate: EffectEstimate;
}

export interface CompareModelsRequest {
  model_ids: string[];
  parse_result?: ParseDemandResponse | Record<string, unknown>;
}

export interface CompareModelsResponse {
  request_id: string;
  items: ModelComparisonItem[];
  matrix: Array<{
    dimension: string;
    values: Record<string, string>;
  }>;
  disclaimer: string;
}

export type FeedbackAction = 'adopt' | 'reject' | 'favorite';

export interface FeedbackRequest {
  request_id: string;
  model_id: string;
  model_name?: string;
  action: FeedbackAction;
  reason?: string;
  scenario?: string;
  parse_result?: ParseDemandResponse | Record<string, unknown>;
  metadata?: Record<string, unknown>;
  evidence_mode?: 'human' | 'demo' | 'test';
}

export interface FeedbackResponse {
  event_id: string;
  status: string;
}

export interface ModelFeedbackStats {
  model_id: string;
  model_name: string;
  scenario: string;
  recommended_count: number;
  adopt_count: number;
  reject_count: number;
  favorite_count: number;
  adoption_rate: number;
}

export interface FeedbackStatsResponse {
  total_events: number;
  items: ModelFeedbackStats[];
  human_event_count: number;
  demo_event_count: number;
  test_event_count: number;
}

export interface RecommendationVersionModel {
  model_id: string;
  model_name: string;
  rank: number;
  total_score: number;
}

export interface RecommendationVersionRecord {
  version_id: string;
  version_number: number;
  session_id: string;
  request_id: string;
  created_at: string;
  parse_summary: {
    intent?: string;
    business_scenario?: string;
    business_stage?: string;
    confidence?: number;
  };
  model_ranking: RecommendationVersionModel[];
  config_hash: string;
}

export interface RecommendationVersionListResponse {
  session_id: string;
  versions: RecommendationVersionRecord[];
  count: number;
}

export interface RecommendationVersionRankChange {
  model_id: string;
  model_name: string;
  rank_a: number;
  rank_b: number;
  rank_delta: number;
  score_a: number;
  score_b: number;
  score_delta: number;
}

export interface RecommendationVersionDiffResponse {
  session_id: string;
  version_a: string;
  version_b: string;
  added_models: string[];
  removed_models: string[];
  rank_changes: RecommendationVersionRankChange[];
  summary: string;
}
